From 30d6d05e1a0e7ceef42bc189a7ba57f846c05ec7 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Fri, 12 Sep 2014 11:48:43 -0400
Subject: [ethernet] i40e: Tell OS link is going down when calling set_phy_config

Message-id: <1410522528-19793-6-git-send-email-sassmann@redhat.com>
Patchwork-id: 93939
O-Subject: [RHEL7.1 PATCH 05/10] i40e: Tell OS link is going down when calling set_phy_config
Bugzilla: 1140994
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

From: Catherine Sullivan <catherine.sullivan@intel.com>

Since we don't seem to be getting an LSE telling us link is going down
during set_phy_config (but we do get an LSE telling us we are coming
back up), fake one for the OS and tell them link is going down. Also
do an atomic restart no matter what because there are times the user
may want to end with link up even if they started with link down (like
if they accidentally set it to a speed that can't link and are trying to
fix it).

Change-ID: I0a642af9c1d0feb67bce741aba1a9c33bd349ed6
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 94128516290dad59d3aaef791faa815c1293298f)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index f0ab3e7..9bc7113 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -559,9 +559,17 @@ static int i40e_set_settings(struct net_device *netdev,
   config.eeer = abilities.eeer_val;
   config.low_power_ctrl = abilities.d3_lpan;
 
-  /* If link is up set link and an so changes take effect */
-  if (hw->phy.link_info.link_info & I40E_AQ_LINK_UP)
-   config.abilities |= I40E_AQ_PHY_ENABLE_ATOMIC_LINK;
+  /* set link and auto negotiation so changes take effect */
+  config.abilities |= I40E_AQ_PHY_ENABLE_ATOMIC_LINK;
+  /* If link is up put link down */
+  if (hw->phy.link_info.link_info & I40E_AQ_LINK_UP) {
+   /* Tell the OS link is going down, the link will go
+    * back up when fw says it is ready asynchronously
+    */
+   netdev_info(netdev, "PHY settings change requested, NIC Link is going down.\n");
+   netif_carrier_off(netdev);
+   netif_tx_stop_all_queues(netdev);
+  }
 
   /* make the aq call */
   status = i40e_aq_set_phy_config(hw, &config, NULL);
@@ -678,6 +686,13 @@ static int i40e_set_pauseparam(struct net_device *netdev,
  else
    return -EINVAL;
 
+ /* Tell the OS link is going down, the link will go back up when fw
+  * says it is ready asynchronously
+  */
+ netdev_info(netdev, "Flow control settings change requested, NIC Link is going down.\n");
+ netif_carrier_off(netdev);
+ netif_tx_stop_all_queues(netdev);
+
  /* Set the fc mode and only restart an if link is up*/
  status = i40e_set_fc(hw, &aq_failures, link_up);
 
-- 
1.7.1